pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1505,7 +1505,7 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
     long retentionMs = TimeUnit.valueOf(validationConfig.getRetentionTimeUnit().toUpperCase())
           .toMillis(Long.parseLong(validationConfig.getRetentionTimeValue()));
     RetentionStrategy retentionStrategy = new TimeRetentionStrategy(TimeUnit.MILLISECONDS,
-          retentionMs - MIN_TIME_BEFORE_SEGMENT_EXPIRATION_FOR_FIXING_DEEP_STORE_COPY_MILLIS);
+        retentionMs - MIN_TIME_BEFORE_SEGMENT_EXPIRATION_FOR_FIXING_DEEP_STORE_COPY_MILLIS, _controllerMetrics);
```

Comment:
Is it desirable behavior to update this metric when isPurgeable returns false when uploading the LLC segment to deepstore?

Reply:
this metric isn't for all instances of isPurgeable=false. It's specifically the case where the zk end time is in an unexpected format, so isPurgeable will always be false. Rather than "this segment just isn't old enough yet".

Reply:
Can we instead have this metric emission from `SegmentStatusChecker#updateSegmentMetrics`? We already fetch segmentZKMetadata over there as well and can easily validate this field.

We can also create a new periodic task - `SegmentZKMetadataValidator` and easily extend it to have multiple validations in future. We can keep retention-flow separate of this metadata-validation imo.

Reply:
The reason I chose not to do it async is because this is the place that's doing the pruning check and failing. If we do it In `SegmentStatusChecker`, then we need to take into account "does this table even have retention enabled". Because that's the only time, as far as I know, that we really care about this case.

I guess I'm looking at this less like a "segment checker" and more like "the retention feature is failing"

Reply:
Having this check in "segment checker" covers "retention feature is failing" plus gives us the scope of extending it to have multiple different checks in future.

Reply:
not exactly. We'd have to duplicate the logic in https://github.com/apache/pinot/blob/master/pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/retention/RetentionManager.java#L95 into the status checker as well. Or make "isRetentionEnabled" a static function and call it.

Either way, seems like we disagree on where to place this. Let's get a third opinion. cc @Jackie-Jiang, curious on your thoughts since you've most recently touched both the `SegmentStatusChecker` and `TimeRetentionStrategy` code